File: lib/koalemos/parsers/css_parser.ex
defmodule Koalemos.Parsers.CSSParser do
  @moduledoc """
  CSS parser for wireframe editor.

  Parses CSS into structured rules for manipulation and display.
  Uses Node.js `css` library for robust parsing.

  ## Usage

      iex> css = ".button { padding: 20px; margin: 10px; }"
      iex> {:ok, result} = CSSParser.parse(css)
      iex> result.rules
      %{".button" => %{"padding" => "20px", "margin" => "10px"}}

  ## Output Format

      %{
        rules: %{
          ".button" => %{"padding" => "20px", "margin" => "10px"},
          "h1, h2" => %{"color" => "blue"}
        }
      }

  ## Notes

  - Ignores @media, @keyframes, and other at-rules for MVP
  - Multiple selectors are joined with ", "
  - Duplicate selectors are merged
  - Invalid CSS returns {:error, reason}
  """

  @doc """
  Parse CSS string into structured rules.

  Returns `{:ok, %{rules: %{...}}}` or `{:error, reason}`.

  ## Examples

      iex> CSSParser.parse(".btn { color: red; }")
      {:ok, %{rules: %{".btn" => %{"color" => "red"}}}}

      iex> CSSParser.parse("invalid {{{")
      {:error, "CSS parsing failed: ..."}

  """
  def parse(css_content) when is_binary(css_content) do
    try do
      result = NodeJS.call!({"css_parser", :parseCSS}, [css_content])
      {:ok, %{rules: Map.get(result, "rules", %{})}}
    rescue
      e ->
        {:error, "CSS parsing failed: #{Exception.message(e)}"}
    end
  end
end

File: priv/nodejs/css_parser.js
/**
 * CSS Parser using `css` library
 *
 * Parses CSS into structured rules for wireframe editor.
 * Similar architecture to js_parser.js.
 *
 * Input: CSS string
 * Output: { rules: { selector: declarations, ... } }
 */

const css = require('css');

/**
 * Parse CSS string into structured rules
 * @param {string} cssContent - The CSS content to parse
 * @returns {object} - { rules: { selector: declarations, ... } }
 */
function parseCSS(cssContent) {
  try {
    const ast = css.parse(cssContent, { silent: false });

    const rules = {};

    // Extract rules from stylesheet
    if (ast.stylesheet && ast.stylesheet.rules) {
      for (const rule of ast.stylesheet.rules) {
        if (rule.type === 'rule') {
          // Standard CSS rule with selectors and declarations
          const ruleData = extractRule(rule);
          if (ruleData) {
            // Merge declarations if selector already exists (handles duplicate selectors)
            if (rules[ruleData.selector]) {
              Object.assign(rules[ruleData.selector], ruleData.declarations);
            } else {
              rules[ruleData.selector] = ruleData.declarations;
            }
          }
        }
        // Ignore @media, @keyframes, etc. for MVP
        // Can add support later if needed
      }
    }

    return { rules };

  } catch (error) {
    throw new Error(`CSS parsing failed: ${error.message}`);
  }
}

/**
 * Extract a single CSS rule
 * @param {object} rule - AST rule node
 * @returns {object} - {selector, declarations}
 */
function extractRule(rule) {
  if (!rule.selectors || rule.selectors.length === 0) {
    return null;
  }

  // Join multiple selectors with comma (e.g., "h1, h2, h3")
  const selector = rule.selectors.join(', ');

  // Extract declarations as key-value pairs
  const declarations = {};

  if (rule.declarations) {
    for (const decl of rule.declarations) {
      if (decl.type === 'declaration' && decl.property && decl.value) {
        declarations[decl.property] = decl.value;
      }
    }
  }

  return {
    selector,
    declarations
  };
}

// Export for use in NodeJS module
module.exports = { parseCSS };

// CLI interface for testing
if (require.main === module) {
  const cssContent = process.argv[2] || '';

  try {
    const result = parseCSS(cssContent);
    console.log(JSON.stringify(result, null, 2));
  } catch (error) {
    console.error(error.message);
    process.exit(1);
  }
}
